include/smp/lock.h

```diff
@@ -22,86 +22,48 @@
 typedef enum {
     CLHState_Granted = 0,
     CLHState_Pending
-} clh_qnode_state_t;
-
-typedef struct clh_qnode {
-    clh_qnode_state_t value;
-
-    PAD_TO_NEXT_CACHE_LN(sizeof(clh_qnode_state_t));
-} clh_qnode_t;
-
-typedef struct clh_qnode_p {
-    clh_qnode_t *node;
-    clh_qnode_t *next;
-    /* This is the software IPI flag */
+} clh_req_state_t;
+
+/* Lock request */
+typedef struct clh_req {
+    clh_req_state_t state;
+} ALIGN(L1_CACHE_LINE_SIZE) clh_req_t;
+
+/* Our node (called "Process" in the paper) */
+typedef struct clh_node {
+    clh_req_t *watch; // Used by predecessor to grant the lock to us.
+    clh_req_t *myreq; // Used to grant the lock to our successor.
+    /* This is the software blocking IPI flag */
     word_t ipi;
-
-    PAD_TO_NEXT_CACHE_LN(sizeof(clh_qnode_t *) +
-                         sizeof(clh_qnode_t *) +
-                         sizeof(word_t));
-} clh_qnode_p_t;
+} ALIGN(L1_CACHE_LINE_SIZE) clh_node_t;
 
 typedef struct clh_lock {
-    clh_qnode_t nodes[CONFIG_MAX_NUM_NODES + 1];
-    clh_qnode_p_t node_owners[CONFIG_MAX_NUM_NODES];
+    clh_req_t request[CONFIG_MAX_NUM_NODES + 1];
+    clh_node_t node[CONFIG_MAX_NUM_NODES];
 
-    clh_qnode_t *head;
-    PAD_TO_NEXT_CACHE_LN(sizeof(clh_qnode_t *));
-} clh_lock_t;
+    clh_req_t *tail;
+} ALIGN(EXCL_RES_GRANULE_SIZE) clh_lock_t;
 
 extern clh_lock_t big_kernel_lock;
 BOOT_CODE void clh_lock_init(void);
 
 static inline bool_t FORCE_INLINE clh_is_ipi_pending(word_t cpu)
 {
-    return big_kernel_lock.node_owners[cpu].ipi == 1;
+    return big_kernel_lock.node[cpu].ipi == 1;
 }
 
-static inline void *sel4_atomic_exchange(void *ptr, bool_t
-                                         irqPath, word_t cpu, int memorder)
+static inline void FORCE_INLINE clh_lock_acquire(bool_t irqPath)
 {
-    clh_qnode_t *prev;
+    word_t cpu = getCurrentCPUIndex();
+    clh_node_t *node = &big_kernel_lock.node[cpu];
```

Comment:
This is generic code, so let's add a save guard that catches development errors with the CPU index mapping.
```
assert(cpu < ARRAY_SIZE(big_kernel_lock.node));
```

Reply:
No one else checks the return value of `getCurrentCPUIndex()`. If that's broken you will never get to an assertion here.

If you want to check it, check `getCurrentCPUIndex() < CONFIG_MAX_NUM_NODES` during boot.